graphbolt/src/cnumpy.cc

```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
   auto result_buffer = reinterpret_cast<char *>(result.data_ptr());
 
   // Indicator for index error.
-  std::atomic<bool> error_flag{};
+  std::atomic<int> error_flag{};
   std::atomic<int64_t> work_queue{};
   torch::parallel_for(0, num_thread_, 1, [&](int64_t begin, int64_t end) {
+    if (begin >= end) return;
     const auto thread_id = begin;
-    // Record the inside offsets of fetched features.
-    int64_t residual[kGroupSize];
-    while (true) {
-      begin = work_queue.fetch_add(kGroupSize, std::memory_order_relaxed);
-      if (begin >= index.numel() || error_flag.load()) break;
-      end = std::min(begin + kGroupSize, index.numel());
+    auto &my_io_uring_queue = io_uring_queue_[thread_id];
+    auto my_read_buffer = ReadBuffer(thread_id);
+    // The completion queue might contain 4 * kGroupSize while we may submit
+    // 4 * kGroupSize more. No harm in overallocation here.
+    CircularQueue<ReadRequest> read_queue(8 * kGroupSize);
+    int64_t num_submitted = 0;
+    int64_t num_completed = 0;
+    auto Submit = [&](int64_t submission_minimum_batch_size) {
+      if (read_queue.Size() < submission_minimum_batch_size) return;
+      TORCH_CHECK(  // Check for sqe overflow.
+          read_queue.Size() <= 2 * kGroupSize);
+      TORCH_CHECK(  // Check for cqe overflow.
+          read_queue.Size() + num_submitted - num_completed <= 4 * kGroupSize);
+      // Submit and wait for the reads.
+      while (!read_queue.IsEmpty()) {
+        const auto submitted = ::io_uring_submit(&my_io_uring_queue);
+        TORCH_CHECK(submitted >= 0);
+        num_submitted += submitted;
+        // Pop the submitted entries from the queue.
+        read_queue.PopN(submitted);
+      }
+    };
+    for (int64_t read_buffer_slot = 0; true;) {
```

Comment:
nit: The whole IndexSelectIOUringImpl has nested many layers, it is better to to breakout to sub-function for better readability.

Reply:
Let me put this to my todo list.